File: backend/src/main/java/org/globe42/domain/IncomeSource.java
package org.globe42.domain;

import java.math.BigDecimal;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.validation.constraints.NotNull;

import org.hibernate.validator.constraints.NotEmpty;

/**
 * A source of income. The same source is shared by many persons, through the {@link Income} entity.
 * @author JB Nizet
 */
@Entity
public class IncomeSource {
    private static final String INCOME_SOURCE_GENERATOR = "IncomeSourceGenerator";

    @Id
    @SequenceGenerator(name = INCOME_SOURCE_GENERATOR, sequenceName = "INCOME_SOURCE_SEQ", initialValue = 1000, allocationSize = 1)
    @GeneratedValue(generator = INCOME_SOURCE_GENERATOR)
    private Long id;

    /**
     * The name of the source of income (APL, Agirc/Arrco, etc.)
     */
    @NotEmpty
    private String name;

    /**
     * The type of the income source
     */
    @ManyToOne
    @NotNull
    private IncomeSourceType type;

    /**
     * The maximum monthly amount (in euros) that the income source can give as income. Null if no known maximum.
     */
    private BigDecimal maxMonthlyAmount;

    public IncomeSource() {
    }

    public IncomeSource(Long id) {
        this.id = id;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public IncomeSourceType getType() {
        return type;
    }

    public void setType(IncomeSourceType type) {
        this.type = type;
    }

    public BigDecimal getMaxMonthlyAmount() {
        return maxMonthlyAmount;
    }

    public void setMaxMonthlyAmount(BigDecimal maxMonthlyAmount) {
        this.maxMonthlyAmount = maxMonthlyAmount;
    }
}

File: backend/src/main/kotlin/org/globe42/domain/IncomeSource.kt
package org.globe42.domain

import java.math.BigDecimal
import javax.persistence.*
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull

private const val INCOME_SOURCE_GENERATOR = "IncomeSourceGenerator"

/**
 * A source of income. The same source is shared by many persons, through the [Income] entity.
 * @author JB Nizet
 */
@Entity
class IncomeSource {

    @Id
    @SequenceGenerator(
        name = INCOME_SOURCE_GENERATOR,
        sequenceName = "INCOME_SOURCE_SEQ",
        initialValue = 1000,
        allocationSize = 1
    )
    @GeneratedValue(generator = INCOME_SOURCE_GENERATOR)
    var id: Long? = null

    /**
     * The name of the source of income (APL, Agirc/Arrco, etc.)
     */
    @NotEmpty
    var name: String? = null

    /**
     * The type of the income source
     */
    @ManyToOne
    @NotNull
    var type: IncomeSourceType? = null

    /**
     * The maximum monthly amount (in euros) that the income source can give as income. Null if no known maximum.
     */
    var maxMonthlyAmount: BigDecimal? = null

    constructor()

    constructor(id: Long) {
        this.id = id
    }
}
